### PR TITLE
Simplify Stack and Runner

### PR DESCRIPTION
**Pull request by pawelprazak**
# Description

This PR focuses on decoupling Stack and Runner, and as a consequence, simplifies both.

Goals:
- decouple Stack from Runner so that Runner knows nothing about Stack
- simplify Runner interface
- maintain "short-circuiting semantics" of the runner (and running the user's code in a context of a stack in general)

Non-goals:
- do not alter significantly other parts of the code, e.g. the testing API, try to stay on target

Changes:
- decouple Stack from Runner
- improve test results API
- make the Stack initialization clear (and gathered in one place)
- simplify and generify the Runner/DefaultRunner
- simplify Stack and move it to an internal package
- remove unused code

I believe we will benefit from those changes when adding features like Stack transformations and Automation API and also when writing internal tests (simpler tests, more things can be tested now, more things can be tested in isolation).

This will also help to polish the test framework.

Part of #626

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: sdk/java/pulumi/src/main/java/com/pulumi/context/internal/ContextInternal.java [hunk 1]
```diff
@@ -30,15 +30,14 @@ public ContextInternal(
             String stackName,
             LoggingContextInternal logging,
             ConfigContextInternal config,
-            OutputContextInternal outputs,
-            Map<String, Output<?>> exports
```
Comment (pawelprazak) [inline sdk/java/pulumi/src/main/java/com/pulumi/context/internal/ContextInternal.java:34]: I've realized that I cannot think of any scenario where this would be pre-set. Please correct me if I'm wrong.

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java [hunk 2]
```diff
@@ -1700,43 +1699,31 @@ public DefaultRunner(Logger standardLogger, EngineLogger engineLogger) {
...
-
-            return whileRunningAsync();
+        public <T> CompletableFuture<Result<T>> runAsync(Supplier<T> callback) {
+            var valueFuture = CompletableFuture.supplyAsync(callback);
+            // run the callback asynchronously in the context of the error handler
+            registerTask("DefaultRunner#runAsync", valueFuture);
+            // loop starts after the callback
+            return valueFuture
+                    .thenCompose(__ -> whileRunningAsync())
+                    .handle((code, throwable) -> {
```
Comment (pawelprazak) [inline sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java:1710]: I wanted to make sure we always handle exceptions.

File: sdk/java/pulumi/src/main/java/com/pulumi/internal/PulumiInternal.java [hunk 3]
```diff
@@ -42,28 +44,35 @@ public static PulumiInternal fromEnvironment() {
...
-                        }
-                )
+    public CompletableFuture<Integer> runAsync(Consumer<Context> stackCallback) {
+        return runAsyncResult(stackCallback).thenApply(r -> r.exitCode());
+    }
+
+    protected CompletableFuture<Result<Stack>> runAsyncResult(Consumer<Context> stackCallback) {
+        // Stack must be created and set globally before running any user code
+        return runner.runAsync(
+                () -> Stack.factory(
```
Comment (pawelprazak) [inline sdk/java/pulumi/src/main/java/com/pulumi/internal/PulumiInternal.java:66]: Wanted to make it clear what is the sequence of events.

File: sdk/java/pulumi/src/main/java/com/pulumi/resources/Resource.java [hunk 4]
```diff
@@ -502,61 +499,6 @@ static public ProviderResource getProvider(Resource resource, String moduleMembe
...
-         * Finds all {@link Output} fields annotated with {@link Export},
-         * validates the annotation presence and content,
-         * validates non-null-ness of fields,
-         * ensures the field type is {@link Output}
-         * and uses reflection to get the references to them.
-         * Returns a map of export names and output references.
-         * Not to be confused with {@link com.pulumi.core.internal.OutputCompletionSource#from(com.pulumi.resources.Resource)}
-         */
-        @InternalUse
-        public static Map<String, Output<?>> findOutputs(Object object) {
```
Comment (pawelprazak) [inline sdk/java/pulumi/src/main/java/com/pulumi/resources/Resource.java:516]: No longer used thanks to previous refactorings.

File: sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+@InternalUse
+@ParametersAreNonnullByDefault
+public final class Stack extends ComponentResource {
```
Comment (pawelprazak) [inline sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java:21]: I wanted to radically simplify this class so there is no magic in the constructor, apart from the inherited magic)

File: sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    public Output<Map<String, Output<?>>> getOutputs() {
+        return this.outputs;
+    }
+
+    @InternalUse
+    public static Function<Supplier<Map<String, Output<?>>>, Stack> factory(
+            String projectName,
+            String stackName,
+            List<ResourceTransformation> resourceTransformations
+    ) {
```
Comment (pawelprazak) [inline sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java:76]: Wanted to make it clear what is the sequence of events and to have the whole init sequence in one place to help with maintaining and debugging in the future.

File: sdk/java/pulumi/src/main/java/com/pulumi/test/internal/PulumiTestInternal.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import com.pulumi.internal.PulumiInternal;
+import com.pulumi.test.PulumiTest;
+import com.pulumi.test.TestResult;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+
+import static java.util.Objects.requireNonNull;
+
+public class PulumiTestInternal extends PulumiInternal implements PulumiTest {
```
Comment (pawelprazak) [inline sdk/java/pulumi/src/main/java/com/pulumi/test/internal/PulumiTestInternal.java:18]: This will get completed in the future.

File: sdk/java/pulumi/src/test/java/com/pulumi/deployment/DeploymentExceptionTest.java [hunk 8]
```diff
@@ -38,14 +38,16 @@ static void cleanup() {
     @Test
     void testUrnFutureDoesNotHangOnException() {
         mock.standardLogger.setLevel(Level.OFF);
-        assertThatThrownBy(() -> mock.runTestAsync(ctx -> {
+        var result = mock.runTestAsync(ctx -> {
```
Comment (pawelprazak) [inline sdk/java/pulumi/src/test/java/com/pulumi/deployment/DeploymentExceptionTest.java:41]: the tests no longer throw before `join`

File: sdk/java/pulumi/src/test/java/com/pulumi/resources/internal/StackTest.java [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+        //noinspection unchecked
+        verify(mock.deployment, times(1))
+                .registerResourceOutputs(any(Resource.class), any(Output.class));
+
+        assertThat(result.outputs()).isEmpty();
+    }
+
+    @Test
+    void testImmediateStackReference() {
```
Comment (pawelprazak) [inline sdk/java/pulumi/src/test/java/com/pulumi/resources/internal/StackTest.java:111]: new test casse

File: sdk/java/pulumi/src/main/java/com/pulumi/context/internal/ContextInternal.java (hunk 1 above)
Comment (t0yv0) [reply sdk/java/pulumi/src/main/java/com/pulumi/context/internal/ContextInternal.java:34]: I'm not sure... at this point I'm leaning to "if it's not tested it does not exist" but we do need to build up some test coverage for how users interact with this.

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java [hunk 10]
```diff
@@ -1700,43 +1699,31 @@ public DefaultRunner(Logger standardLogger, EngineLogger engineLogger) {
...
+                        if (throwable != null) {
+                            return handleExceptionAsync(throwable).thenApply(errorCode ->
+                                    new Result<>(
+                                            errorCode,
+                                            ImmutableList.copyOf(this.swallowedExceptions),
+                                            Optional.ofNullable(valueFuture.join()) // the future is already complete
```
[the remark this answers is not in the document]
Comment (t0yv0) [reply sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java:1716]: It completed with an exception though, why do you think it would have a value? Naively I'd expect `.join()` to throw here but I'm not sure, better try it out and see.

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java (hunk 2 above)
Comment (t0yv0) [reply sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java:1710]: Doesn't registerTask handle exceptions also? 

This might be ok but a bit convoluted to reason about.

Comment (t0yv0) [reply sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java:1710]: Let me check where runAsync is called.

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java [hunk 11]
```diff
@@ -1700,43 +1699,31 @@ public DefaultRunner(Logger standardLogger, EngineLogger engineLogger) {
...
+                                    )
+                            );
+                        }
+                        return CompletableFuture.completedFuture(new Result<>(
+                                code,
+                                ImmutableList.copyOf(this.swallowedExceptions),
+                                Optional.ofNullable(valueFuture.join()) // the future is already complete
```
[the remark this answers is not in the document]
Comment (t0yv0) [reply sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java:1723]: Here, yes.

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java (hunk 2 above)
Comment (t0yv0) [reply sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java:1710]: OK. This implements the project entry point. I think this isn't right. This code makes the `CompletableFuture<Result<T>>` ALWAYS wait for `valueFuture` which removes the feature of early termination due to an out-of-band exception somewhere in the underlying net of futures. That is, our protocol up to now, was that if any exception happens anywhere in the Output chain, we bubble up and stop. This is the point of registerTask and whileRunningAsync.

Is this an intentional change to the semantics?

I wonder what happened with the test that tested for it.

File: sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    @InternalUse
+    public static final String RootPulumiStackTypeName = "pulumi:pulumi:Stack";
+
+    /**
+     * The outputs of this stack, if the <code>callback</code> callback exited normally.
+     */
+    private final Output<Map<String, Output<?>>> outputs;
+
+    /**
+     * Create a Stack with stack outputs defined in derived class constructor.
```
Comment (t0yv0) [inline sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java:37]: The comment is weird, what is "derived class constructor"? Stack is now final right?

File: sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+        // - set the Stack instance in the global state (user code will try to reach it via Resource constructor)
+        // - register the Stack outputs in the Pulumi engine
+        // - call the user code callback (asynchronously to avoid any additional cyclic dependency problems)
+        // - complete the Stack initialization (by completing the lazyFuture)
+        return (Supplier<Map<String, Output<?>>> callback) -> {
+            // incomplete future that will be completed after the Stack is constructed
+            var lazyFuture = new CompletableFuture<Map<String, Output<?>>>();
+            // incomplete outputs that will be completed when lazyFuture completes
+            var lazyOutputs = Output.of(lazyFuture);
+            // Stack must be the first Resource to be constructed,
```
Comment (t0yv0) [inline sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java:88]: Not a fan of commenting every line, that's diminishing returns for comments the code speaks for itself ideally at this level.

File: sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+            // before any user code attempts to construct any Resource
+            var stack = new Stack(projectName, stackName, resourceTransformations, lazyOutputs);
```
Comment (t0yv0) [inline sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java:90]: I have some doubts this sequencing works, I worry about a race. Does `new Stack()` launch `RegisterResource` or some other async call in a bg thread? Or it doesn't because it's a Component? Need to read it carefully. 

It seems that it *does* call RegisterResource to the ResourceMonitor gRPC service.

Now that will race with `deploymentInstance.setStack(stack)` and  `DeploymentInternal.getInstance().registerResourceOutputs(stack, lazyOutputs);`. This may or may not be okay? Would something call `deploymentInstance.getStack()` before the RegisterResource thing?

File: sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+            // register the new Stack with the engine
+            DeploymentInternal.getInstance().setStack(stack);
+            // register the Stack outputs with the engine
+            DeploymentInternal.getInstance().registerResourceOutputs(stack, lazyOutputs);
+            // run the user code callback wrapped in a future (note that future is eager)
+            var callbackFuture = CompletableFuture.supplyAsync(callback);
+            DeploymentInternal.getInstance().getRunner().registerTask("callback", callbackFuture);
```
Comment (t0yv0) [inline sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java:97]: Perhaps share `DeploymentInternal.getInstance()` or even make it a param so we query the global less.

File: sdk/java/pulumi/src/main/java/com/pulumi/test/TestResult.java [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * A test result information.
+ */
+public class TestResult {
+    private final int exitCode;
+    private final List<Resource> resources;
+    private final List<Exception> exceptions;
+    private final List<String> errors;
```
Comment (t0yv0) [inline sdk/java/pulumi/src/main/java/com/pulumi/test/TestResult.java:21]: What's the difference between errors and exceptions?

File: sdk/java/pulumi/src/main/java/com/pulumi/test/TestResult.java [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+/**
+ * A test result information.
+ */
+public class TestResult {
+    private final int exitCode;
```
Comment (t0yv0) [inline sdk/java/pulumi/src/main/java/com/pulumi/test/TestResult.java:18]: Why do we need to test exit codes?

File: sdk/java/pulumi/src/test/java/com/pulumi/deployment/DeploymentExceptionTest.java (hunk 8 above)
Comment (t0yv0) [reply sdk/java/pulumi/src/test/java/com/pulumi/deployment/DeploymentExceptionTest.java:40]: That seems actually good. Consistent.

**Comment by t0yv0** [comment]
This is mostly good, as usual it would benefit from breaking down into smaller PRs but we can review as is.

1. I have some doubts about early termination in Runner that might be blocking the PR for me above; can we talk through that? You believe it works as expected? 

2. Ideally we need CHANGELOG_PENDING entry outlining all the changes to user-facing functionality; it's hard for me to even see which one is that, but at the very least: explain that Stack class is removed and why, and explain that the mock testing interface is changed and how.

File: sdk/java/pulumi/src/main/java/com/pulumi/context/internal/ContextInternal.java (hunk 1 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/context/internal/ContextInternal.java:34]: > "if it's not tested it does not exist"

I like this. Since this is an internal API, this constructor, then it can be easily added back :)

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/MockRunner.java [hunk 18]
```diff
@@ -1,39 +1,19 @@
...
-    }
-
-    @Override
-    public CompletableFuture<Integer> runAsyncFuture(Supplier<CompletableFuture<Map<String, Output<?>>>> callback, StackOptions options) {
-        return null;
-    }
-
-    @Override
-    public <T extends Stack> CompletableFuture<Integer> runAsync(Supplier<T> stackFactory) {
+    public <T> CompletableFuture<Result<T>> runAsync(Supplier<T> callback) {
```
[the remark this answers is not in the document]
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/deployment/MockRunner.java:16]: > Why is this class public and marked internal etc? Do you see users interacting with it at any point?
I wanted to move it to an internal package. I see no reason for it to be a public API. (ofc the public keyword will have to be there to be visible from other internal packages.

I didn't want to make too many changes in too many places at once. I plan to address the testing API in #553

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java:1716]: Hmm, that's a good point, does it throw, it should throw on `join`.
I can't exclude a situation where both the value and an exception is present, but not sure if it is a useful state. 
Maybe simplifying this to always return Optional.empty() could be a better option, but I was worried this will make some tests impossible/difficult to write.

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java (hunk 2 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java:1710]: > Is this an intentional change to the semantics?
No, it is not intentional. I was under the impression, this maintains the semantics, but I'm not sure anymore.

> I wonder what happened with the test that tested for it.
I'd guess it either does not test this conditions well enough or the semantics didn't actually changed.

I was under the impression that `valueFuture` would short circuit on an exception and not even call the `.thenCompose(__ -> whileRunningAsync())` and go straight to `handle`.

Let me try to come up with a test for those short-circuiting semantincs.

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/Runner.java [hunk 19]
```diff
@@ -1,14 +1,18 @@
...
 import java.util.concurrent.CompletableFuture;
 import java.util.function.Supplier;
 
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A task {@link CompletableFuture} runner.
+ */
+@InternalUse
 public interface Runner {
```
[the remark this answers is not in the document]
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/Runner.java:16]: > 1. why does it deal with exit codes at all - at this level?
yes, this is confusing for me as well, I was thinking about moving the error handling out of the runner, or at least out of the mian loop and do it once with a `.handle` not multiple times

> 2. why does the caller need to get swallowed exceptions, [...]
I've removed `getSwallowedExceptions` and have put the exceptions into the `Result`. Right now it is needed for the tests only. Not sure if universally useful. The functionality was not originally there, I've ported it from C# while trying to catch up with upstream changes.

> 4. [...]
I'd love to be able to simplify the signature even more: `CompletableFuture<Void> run()` I'm just not sure if we can guarantee the stack will be created before user code is called just by registering both and making the user's code execution lazy/wait for stack

Then we could make the error handling on the boundary in PulumiInternal, instead of the Runner. 
This way a Runner would do just one thing and be super easy to test and reason about.

I just have a feeling it's not gonna be so easy, but we might try anyway :D

File: sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java (hunk 13 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java:88]: I agree, ideally code should speak for itself. After struggling to understand the sequence and the reasons behind it myself I wanted to help everyone else who might need to do the same in the future.
Would rewriting the comments into a one big description on to be more readable for you?

File: sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java (hunk 14 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java:90]: Great points. My understanding is that `getStack()` will only be called for all other resources then the `Stack resource.

Currently there are two places where `getStack` is called:
- (read or register resource) `prepareResourceAsync` -> `getRootResourceAsync` - if a resource has no parent and is not a Stack
- Resource constructor - if a resource has no parent and is not a Stack

File: sdk/java/pulumi/src/main/java/com/pulumi/test/TestResult.java (hunk 16 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/test/TestResult.java:21]: errors are error messages from logger, exceptions are exception instances

File: sdk/java/pulumi/src/test/java/com/pulumi/deployment/MocksTest.java [hunk 20]
```diff
@@ -118,7 +118,7 @@ void testStack() {
                 .build();
 
         var result = mock.runTestAsync(MocksTest::myStack).join();
-        var publicIp = waitForValue(result.stackOutput("publicIp", String.class));
```
[the remark this answers is not in the document]
Comment (pawelprazak) [reply sdk/java/pulumi/src/test/java/com/pulumi/deployment/MocksTest.java:121]: if we want to resonate with `pulumi stack output` a natural thing IMO would be to have:

```
PulumiTest.run(ctx -> ...)
    .stack.output("foo", String.class);
```

So `Stack` in the Result itself.

**Comment by pawelprazak** [comment]
> smaller PRs but we can review as is

Sorry about the volume of changes, I couldn't make smaller PRs to make them sensible - they would appear rather random I'm afraid without the context.

Maybe stacking multiple smaller PR would improve things? Although not convinced about the benefits of smaller PR vs having multiple commits in a PR, eps. when untangling things, they tend to touch many places in the code.

I'll keep that in mind thou, maybe next time I'll find a better way to make a PR smaller and easier to review.

File: sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java (hunk 14 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java:90]: Regarding outputs registration, AFAIU, `lazyOutputs` being not completed will block the registration until completed via `whenComplete`.

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java (hunk 2 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java:1710]: After adding more test cases I was not successful in triggering the `if (throwable != null)` branch.

It turns out it might be redundant with the code in the loop itself. I wonder if we could remove the handler form the loop and handle exceptions here instead maintaining the semantics.

File: sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java (hunk 15 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java:123]: done

File: sdk/java/pulumi/src/main/java/com/pulumi/test/TestResult.java (hunk 16 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/test/TestResult.java:21]: we terminate on a logged error even if there were no exceptions (behaviour inherited from C# impl)

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java]: After moving error handling to one place, the join is no longer needed. I'm glad this small hack of mine got deleted :)

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java (hunk 2 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java]: Yes, after moving the handler to only be present in `handle` all tests pass and it looks equivalent (in behaviour) to me looking at the code.

File: sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java (hunk 12 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java]: fixed

File: sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java (hunk 13 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java]: I've moved the comments/clarify, hope it helps.

File: sdk/java/pulumi/src/main/java/com/pulumi/test/TestResult.java (hunk 17 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/test/TestResult.java:18]: Yes, after giving this a though, exit codes are special case that works only for the CLI use and are not very useful to the user anyway. Moving the exit codes out of runner and removing from test result seam like a good idea.
Thank you for spotting this inconsistency.

File: sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java (hunk 14 above)
Comment (t0yv0) [reply sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java:120]: Good! In that case we should be OK here!

File: sdk/java/pulumi/src/main/java/com/pulumi/test/TestResult.java (hunk 16 above)
Comment (t0yv0) [reply sdk/java/pulumi/src/main/java/com/pulumi/test/TestResult.java:21]: Gotcha, so this basically captures what was logged.

File: sdk/java/pulumi/src/test/java/com/pulumi/deployment/MocksTest.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (t0yv0) [reply sdk/java/pulumi/src/test/java/com/pulumi/deployment/MocksTest.java:121]: I like it. Good mnemonic!

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java [hunk 21]
```diff
@@ -1706,21 +1706,25 @@ public <T> CompletableFuture<Result<T>> runAsync(Supplier<T> callback) {
             registerTask("DefaultRunner#runAsync", valueFuture);
             // loop starts after the callback
             return valueFuture
-                    .thenCompose(__ -> whileRunningAsync())
-                    .handle((code, throwable) -> {
+                    .thenCompose(value -> whileRunningAsync().thenApply(__ -> value))
```
Comment (t0yv0) [inline sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java:1709]: Why not this:

```
whileRunningAsync().thenCompose(_ -> valueFuture)
```

Or use `thenCombine` API.

What I am trying to see is that when `valueFuture` takes a long time, but there's an exception out of band, this exception short-circuits `valueFuture` instead of waiting on it. Let me try to make a test.

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/Runner.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/Runner.java:16]: I agree, I've gathered the feedback regarding possible future directions in #668 for further considerations.

File: sdk/java/pulumi/src/test/java/com/pulumi/deployment/MocksTest.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (pawelprazak) [reply sdk/java/pulumi/src/test/java/com/pulumi/deployment/MocksTest.java:121]: Added to #668 for reference.

File: sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java (hunk 21 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java:1709]: I'm slightly inclined to leave as-is, because this looks more confusing in my opinion:
`valueFuture.thenCompose(__ -> whileRunningAsync().thenCompose(___ -> valueFuture))` 
than this
`valueFuture.thenCompose(value -> whileRunningAsync().thenApply(__ -> value))`

Unless I'm missing something, they are equivalent in behaviour, right? (Tests are green in both cases)
